### PR TITLE
Bs#139: improve browserstack test runner process

### PR DESCRIPTION
**Pull request by happyhj**
# Detail
* chore(all): improve test runner process with 2 browserstack vm  …
    * with "grunt browserstack" command, test will run by browser using 1 vm once.

    screen shot 2016-02-23 at 2 00 59 pm
    * with "grunt browserstack:modulename" command, test will run by target module using 2 vm once.

    screen shot 2016-02-23 at 2 29 57 pm

* test(visible): Run test with iScroll only when iScroll works for old IE (IE 7, 8)

# Issue
#137 #139 

# Preferred reviewer
@netil @sculove

### CODE REVIEW

File: config/browserstack.js [hunk 1]
```diff
@@ -52,6 +52,47 @@ module.exports = function(grunt){
 	    })
 	    return results
 	}
+	
+	function isSameBrowser(browser1, browser2) {
+		if(JSON.stringify(browser1) === JSON.stringify(browser2)) {
+			return true;
```
Comment (netil) [inline config/browserstack.js]:
```js
return JSON.stringify(browser1) === JSON.stringify(browser2);
```

File: config/browserstack.js [hunk 2]
```diff
@@ -77,57 +118,90 @@ module.exports = function(grunt){
...
 			"test_path": [
 			"test/unit/"+componentName+".test.html"
 			],
 			"browsers": targetBrowsers
 		};
-	}
+	}	
 
+	grunt.registerTask('browserstack_runner', function() {
+		var isByBrowser = arguments[0] === "byBrowser" ? true : false;
```
Comment (netil) [inline config/browserstack.js]:
```js
var isByBrowser = arguments[0] === "byBrowser";
```

File: config/browserstack.js [hunk 3]
```diff
@@ -77,57 +118,90 @@ module.exports = function(grunt){
...
+		var browserstackConfig;
+		if(!isByBrowser) {
+			browserstackConfig = getConfig(arguments[0]);
+			var bIdx = 2 * (parseInt(arguments[1]) + 1) - 2;
+			var totalBrowser = browserstackConfig.browsers.length;
+			var browsers = browserstackConfig.browsers;
+			var targetBrowsers = [];
+			targetBrowsers.push(browsers[bIdx]);
```
Comment (netil) [inline config/browserstack.js]:
```js
var targetBrowsers = [ browsers[bIdx] ];
```

File: config/browserstack.js [hunk 4]
```diff
@@ -77,57 +118,90 @@ module.exports = function(grunt){
...
+	function runBrowserstackRunner(browserstackConfig) {
+		var tempBrowserstackConfig = "config/browertstack.config.json";
+		process.env["BROWSERSTACK_JSON"] = tempBrowserstackConfig;
+		fs.writeFileSync(tempBrowserstackConfig, JSON.stringify(browserstackConfig), "utf8");
+	
+		var done = this.async();
+		
+		var subProcess = exec("node_modules/.bin/browserstack-runner --verbose", function(err) {
+			var fs = require("fs");
+			done(err ? false : true);
```
Comment (netil) [inline config/browserstack.js]:
```js
done(!err);
```

File: config/browserstack.js [hunk 5]
```diff
@@ -52,6 +52,47 @@ module.exports = function(grunt){
...
+	function getConfigByBrowser() {
+		// moduleList getConfig
+		var configList = moduleList.map(function(val) {
+			return getConfig(val.name);
+		});
+		
+		var browsers = [];
+		for(var i in bsLaunchers) {
+			browsers = browsers.concat(bsLaunchers[i]);
+		}
```
Comment (mixed) [inline config/browserstack.js]: Could you use [`reduce`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce)?
```
var browsers = variable.reduce(function(a,b){
 return a.concat(b);
},[]);
```

File: config/browserstack.js [hunk 6]
```diff
@@ -52,6 +52,47 @@ module.exports = function(grunt){
...
+		var configListByBrowser = browsers.map(function(val){
+			return { 
+				test_framework: 'qunit',
+				test_path: [],
+				browsers: [val]
+			}
+		});
+		
+		configListByBrowser.forEach(function(config) {
```
Comment (mixed) [inline config/browserstack.js]: IMO This logic is `map` better.

File: config/browserstack.js [hunk 7]
```diff
@@ -52,6 +52,47 @@ module.exports = function(grunt){
...
+		return false;
+	}
+	
+	function hasBrowser(browserList, browser) {
+		return browserList.some(function(val){
+			return isSameBrowser(val, browser);
+		});
+	}
+	
+	function getConfigByBrowser() {
```
Comment (mixed) [inline config/browserstack.js]: IMO I prefer to none variable. So. Could you use `method chaining`?
```
var configList =...;
return bsLaunchers.reduce
                     .map
                     .map
```

File: config/browserstack.js [hunk 8]
```diff
@@ -52,6 +52,47 @@ module.exports = function(grunt){
...
+		// moduleList getConfig
+		var configList = moduleList.map(function(val) {
```
Comment (mixed) [inline config/browserstack.js]: IMO I prefer `closure` to `parameter`.

File: config/browserstack.js (hunk 5 above)
Comment (happyhj) [inline config/browserstack.js]: bsLaunchers is not an Array, so can not use reduce method.

**Comment by mixed** [comment]
LGTM

**Comment by netil** [comment]
LGTM

**Pull request closed without merging**